lib/shape.ml

```diff
@@ -96,52 +112,92 @@ let scale factor s =
         (scale_length ellipse'.ry factor)
   | Line _line' -> failwith "Not Implemented"
 
+(* Show a list of shapes on the canvas *)
 let show shapes = List.iter render_shape shapes
 
-let bi_to_uni x y =
-  let nx = (x *. 0.5) +. (float_of_int !dimensions.x *. 0.5) in
-  let ny = (y *. 0.5) +. (float_of_int !dimensions.y *. 0.5) in
-  (int_of_float nx, int_of_float ny)
-
+(* Convert degrees to radians *)
 let deg_to_rad degrees = degrees *. (Stdlib.Float.pi /. 180.)
 
-let rot { x : int; y : int } degrees =
-  let radians = deg_to_rad (float_of_int degrees) in
-  let dx = (float_of_int x *. cos radians) -. (float_of_int y *. sin radians) in
-  let dy = (float_of_int x *. sin radians) +. (float_of_int y *. cos radians) in
-  let dx, dy = bi_to_uni dx dy in
-  { x = dx; y = dy }
-
+(* Rotate a shape by a specified number of degrees *)
 let rotate degrees shape =
+  let radians = deg_to_rad degrees in
+
+  (* Convert a point from Cartesian to polar coordinates *)
+  let cartesian_to_polar point =
+    let r = sqrt (float_of_int (point.x * point.x + point.y * point.y)) in
+    let theta = atan2 (float_of_int point.y) (float_of_int point.x) in
+    (r, theta)
+  in
+
+  (* Convert a point from polar to Cartesian coordinates *)
+  let polar_to_cartesian (r, theta) =
+    let x = int_of_float (r *. cos theta) in
+    let y = int_of_float (r *. sin theta) in
+    { x; y }
+  in
+
   match shape with
-  | Circle circle -> Circle { c = rot circle.c degrees; radius = circle.radius }
+  | Circle circle ->
+      let polar_center = cartesian_to_polar (denormalize circle.c) in
+      let polar_rotated = (fst polar_center, snd polar_center +. radians) in
+      let rotated_center = polar_to_cartesian polar_rotated in
+      (* Make sure the rotated circle remains within the canvas *)
+      let max_x = !dimensions.x / 2 in
+      let max_y = !dimensions.y / 2 in
+      let clamped_x = max (min rotated_center.x max_x) (-max_x) in
+      let clamped_y = max (min rotated_center.y max_y) (-max_y) in
```

Comment:
I don't think we need this.

We shouldn't artificially bring in elements that would be out of canvas.

What I meant in the issue was that because we don't do rotation properly, our shapes fly out of the canvas.